File: scripts/update_readme_article_count.py
#!/usr/bin/env python3
"""Synchronize README article counts with the number of markdown files in articles/."""

from pathlib import Path
import re


def get_article_count() -> int:
    articles_dir = Path("articles")
    if not articles_dir.exists():
        raise FileNotFoundError(f"Articles directory not found: {articles_dir}")
    return len(list(articles_dir.glob("*.md")))


def update_readme(count: int) -> bool:
    readme_path = Path("README.md")
    if not readme_path.exists():
        raise FileNotFoundError(f"README not found: {readme_path}")

    count_with_commas = f"{count:,}"
    original = readme_path.read_text(encoding="utf-8")
    updated = original

    updated = re.sub(
        r"(The site contains \*\*)([\d,]+)(\s+articles\*\*)",
        lambda match: f"{match.group(1)}{count_with_commas}{match.group(3)}",
        updated,
    )

    updated = re.sub(
        r"(articles/\s+#\s*)([\d,]+)(\s+Markdown articles)",
        lambda match: f"{match.group(1)}{count}{match.group(3)}",
        updated,
    )

    if updated == original:
        return False

    readme_path.write_text(updated, encoding="utf-8")
    return True


def main() -> int:
    count = get_article_count()
    changed = update_readme(count)
    if changed:
        print(f"Updated README article count to {count}")
    else:
        print(f"README already at {count}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/update_readme_article_count.js
#!/usr/bin/env node
const fs = require('fs');
const path = require('path');

const root = process.cwd();
const articlesDir = path.join(root, 'articles');
const readmePath = path.join(root, 'README.md');

if (!fs.existsSync(articlesDir) || !fs.existsSync(readmePath)) {
  process.exit(1);
}

const articleCount = fs
  .readdirSync(articlesDir)
  .filter((fileName) => fileName.endsWith('.md')).length;
const articleCountWithCommas = articleCount.toLocaleString('en-US');

const original = fs.readFileSync(readmePath, 'utf8');

let updated = original.replace(
  /(The site contains \*\*)([\d,]+)(\s+articles\*\*)/,
  `$1${articleCountWithCommas}$3`
);

updated = updated.replace(
  /(articles\/\s+#\s*)([\d,]+)(\s+Markdown articles)/,
  `$1${articleCount}$3`
);

if (updated !== original) {
  fs.writeFileSync(readmePath, updated, 'utf8');
  process.stdout.write(`Updated README article count to ${articleCount}\n`);
} else {
  process.stdout.write(`README already at ${articleCount}\n`);
}
